bench/clblas_benchmark.cpp

```diff
@@ -0,0 +1,392 @@
+    static constexpr const auto func = &clblasD##postfix; \
+  };
+/* template <> struct clblasX##postfix<std::complex<float>> { \ */
+/*   static constexpr const auto value = clblasC##postfix; \ */
+/* }; \ */
+/* template <> struct clblasX##postfix<std::complex<double>> { \ */
+/*   static constexpr const auto value = clblasZ##postfix; \ */
+/* }; */
+
+#define CLBLASI_FUNCTION(postfix)                          \
+  template <typename T>                                    \
+  struct clblasiX##postfix;                                \
+  template <>                                              \
+  struct clblasiX##postfix<float> {                        \
+    static constexpr const auto func = &clblasiS##postfix; \
+  };                                                       \
+  template <>                                              \
+  struct clblasiX##postfix<double> {                       \
+    static constexpr const auto func = &clblasiD##postfix; \
+  };
+/* template <> struct clblasiX##postfix<std::complex<float>> { \ */
+/*   static constexpr const auto value = clblasiC##postfix; \ */
+/* }; \ */
+/* template <> struct clblasiX##postfix<std::complex<double>> { \ */
+/*   static constexpr const auto value = clblasiZ##postfix; \ */
+/* }; */
+
+CLBLAS_FUNCTION(scal);
+CLBLAS_FUNCTION(asum);
+CLBLAS_FUNCTION(axpy);
+CLBLAS_FUNCTION(nrm2);
+CLBLAS_FUNCTION(dot);
+CLBLASI_FUNCTION(amax);
+
+class ClBlasBenchmarker {
+  Context context;
+
+ public:
+  ClBlasBenchmarker() : context() { clblasSetup(); }
+
+  BENCHMARK_FUNCTION(scal_bench) {
+    UNPACK_PARAM;
+    double flops;
+    {
+      ScalarT alpha(2.4367453465);
+      MemBuffer<ScalarT> buf1(context, size);
+      cl_event event;
+      flops = benchmark<>::measure(no_reps, size * 1, [&]() {
+        clblasXscal<ScalarT>::func(size, alpha, buf1.dev(), 0, 1, 1,
+                                   context._queue(), 0, NULL, &event);
+        clWaitForEvents(1, &event);
+        clReleaseEvent(event);
+      });
+    }
+    return flops;
+  }
+
+  BENCHMARK_FUNCTION(axpy_bench) {
+    UNPACK_PARAM;
+    double flops;
+    {
+      ScalarT alpha(2.4367453465);
+      MemBuffer<ScalarT> buf1(context, size);
+      MemBuffer<ScalarT> buf2(context, size);
+      cl_event event;
+      flops = benchmark<>::measure(no_reps, size * 1, [&]() {
+        clblasXaxpy<ScalarT>::func(size, alpha, buf1.dev(), 0, 1, buf2.dev(), 0,
+                                   1, 1, context._queue(), 0, NULL, &event);
+        clWaitForEvents(1, &event);
+        clReleaseEvent(event);
+      });
+    }
+    return flops;
+  }
+
+  BENCHMARK_FUNCTION(asum_bench) {
+    UNPACK_PARAM;
+    double flops;
+    {
+      ScalarT vr;
+      MemBuffer<ScalarT, CL_MEM_WRITE_ONLY> buf1(context, size);
+      MemBuffer<ScalarT, CL_MEM_READ_ONLY> bufr(context, &vr, 1);
+      cl_mem scratch = clCreateBuffer(context, CL_MEM_READ_WRITE,
+                                      size * sizeof(ScalarT), NULL, NULL);
```

Comment:
Please be aware of the different flags for creating buffers: https://www.khronos.org/registry/OpenCL/sdk/1.2/docs/man/xhtml/clCreateBuffer.html